app/src/main/java/com/zeapo/pwdstore/Application.kt

```diff
@@ -4,52 +4,116 @@
+class Application : android.app.Application(), SharedPreferences.OnSharedPreferenceChangeListener, LifecycleObserver {
 
-    private var prefs: SharedPreferences? = null
+    private lateinit var prefs: SharedPreferences
+    private val coroutineScope = CoroutineScope(Dispatchers.IO)
+    private var authTimeout = 15
+    private var job: Job? = null
+    @Volatile var isAuthEnabled = false
+    @Volatile var requiresAuthentication = true
 
     override fun onCreate() {
         super.onCreate()
         prefs = PreferenceManager.getDefaultSharedPreferences(this)
-        if (BuildConfig.ENABLE_DEBUG_FEATURES || prefs?.getBoolean(PreferenceKeys.ENABLE_DEBUG_LOGGING, false) ==
-            true) {
+        if (BuildConfig.ENABLE_DEBUG_FEATURES || prefs.getBoolean(PreferenceKeys.ENABLE_DEBUG_LOGGING, false)) {
             plant(DebugTree())
         }
-        prefs?.registerOnSharedPreferenceChangeListener(this)
+        prefs.registerOnSharedPreferenceChangeListener(this)
+        isAuthEnabled = prefs.getBoolean(PreferenceKeys.BIOMETRIC_AUTH, false)
+        if (isAuthEnabled) updateAuthTimeout()
         setNightMode()
         setUpBouncyCastleForSshj()
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
     }
 
     override fun onTerminate() {
-        prefs?.unregisterOnSharedPreferenceChangeListener(this)
+        prefs.unregisterOnSharedPreferenceChangeListener(this)
         super.onTerminate()
     }
 
     override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String) {
-        if (key == PreferenceKeys.APP_THEME) {
-            setNightMode()
+        when (key) {
+            PreferenceKeys.APP_THEME -> setNightMode()
+            PreferenceKeys.BIOMETRIC_AUTH_TIMEOUT -> updateAuthTimeout()
+            PreferenceKeys.BIOMETRIC_AUTH -> setAuthentication()
         }
     }
 
     private fun setNightMode() {
-        AppCompatDelegate.setDefaultNightMode(when (prefs?.getString(PreferenceKeys.APP_THEME, getString(R.string.app_theme_def))) {
+        AppCompatDelegate.setDefaultNightMode(when (prefs.getString(PreferenceKeys.APP_THEME, getString(R.string.app_theme_def))) {
             "light" -> MODE_NIGHT_NO
             "dark" -> MODE_NIGHT_YES
             "follow_system" -> MODE_NIGHT_FOLLOW_SYSTEM
             else -> MODE_NIGHT_AUTO_BATTERY
         })
     }
+
+    private fun setAuthentication() {
+        isAuthEnabled = prefs.getBoolean(PreferenceKeys.BIOMETRIC_AUTH, false)
+    }
+
+    private fun updateAuthTimeout() {
+        authTimeout = try {
+            Integer.parseInt(prefs.getString(PreferenceKeys.BIOMETRIC_AUTH_TIMEOUT, "15") as String)
+        } catch (e: NumberFormatException) {
+            15
+        }
+        d { "Auth timeout update : $authTimeout" }
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
+    private fun onResume() {
+        job?.cancel()
+        if (isAuthEnabled && requiresAuthentication) {
```

Comment:
should we be synchronizing accesses to `requiresAuthentication`?

Reply:
I think `@Volatile` would be enough since it makes sure that all read and write operations are atomic.